Connect account webhooks

arnav13081994 opened the pull request:
This PR contains the following changes:

1) Added support for `account.updated` webhook events for events on `Platform Accounts` as well as `Stripe Connected Accounts`. Along with tests verifying functionality for the same. Manually tested on Stripe `Standard`, `Custom` and `Express` Connected Accounts as well.
1) Added support for `account.external_account.*` webhook events for events on Stripe Connected Accounts.  Along with tests verifying functionality for the same. Manually tested on Stripe `Custom` and `Express` Connected Accounts as well.
1) Added support for `account.application.*` webhook events for events on Stripe Connected Accounts.  Along with tests verifying functionality for the same. Manually tested on Stripe `Standard`, `Custom` and `Express` Connected Accounts as well.
1) Updated the management command `djstripe_sync_models` to be able to sync `BankAccount` and `Card` objects for Stripe `Custom` and `Express` Connected Accounts.
1) Updated `admin` to allow a user to be able to checkout `BankAccount` and `Card` objects on `django admin`
1) Added model field `account` to the `Card` model. `Custom` Stripe Connected Accounts can specify a `Card` object as an `External Account` Payout Source.
1) Updated `payment_methods` module to be able to handle the corner case that a marketplace allows a user to function as both a `buyer` and a `seller` using the same `Card` and/or `Bank Account` as a `Payment Source` and an `External Account` Payout Source.
1) Updated `payment_methods` module to be able to `retrieve`, `create` and `list`, `Card` and `BankAccount` objects both by `customer` and/or `account` depending on the case. (_Were Pending TODOs anyway_)
1) Added test fixtures for Stripe `Standard`, `Custom` and `Express` Connected Accounts and removed redundant fixtures and/or dicts.
1) Replaced `default_account` with a Stripe `Standard` account fixture.

**Overall this PR should allow `dj-stripe` to handle all Stripe and Stripe Connected Account Webhook events!**

arnav13081994 committed:
Exposed Card and BankAccount objects on the admin interface.

jleclanche commented on djstripe/event_handlers.py:
```diff
@@ -205,6 +207,8 @@ def other_object_webhook_handler(event):
     else:
         target_cls = {
+            "account": models.Account,
+            # TODO: Add an Application model to handle account.application.auth or de-auth events
```
What happens if you receive such an event right now? Will it crash?

arnav13081994 replied:
Outdated. Also in the final code, I am handling `account.updated` and `account.external_account.*` webhooks event in their own webhook handlers. That makes it much easier to handle and modify the behaviour (if need be) in the future.

jleclanche commented on djstripe/models/payment_methods.py:
```diff
@@ -157,20 +191,35 @@ def api_retrieve(self, api_key=None, stripe_account=None):
-        # so we don't have to pass `stripe_account`.
-        return customer.sources.retrieve(self.id, expand=self.expand_fields)
+            # This will retrieve the external_accounts using the bank_account ID where the account resides,
+            # so we don't have to pass `stripe_account`.
```
Is dropping expand_fields going to be a problem here?

arnav13081994 replied:
I didn't write this line: ```customer.sources.retrieve(self.id, expand=self.expand_fields)``` But looking at the code, I think ```expand=self.expand_fields``` can be dropped without any problem. This is because ```self.customer.api_retrieve(
                api_key=api_key, stripe_account=stripe_account
            )``` will use `expand_fields` anyway.

jleclanche commented on djstripe/models/payment_methods.py:
```diff
@@ -181,7 +230,7 @@ class BankAccount(LegacySourceMixin, StripeModel):
         null=True,
         blank=True,
-        related_name="bank_account",
+        related_name="bank_accounts",
```
This is backwards incompatible, can you add a .bank_account alias property?

arnav13081994 replied:
Ok sure. I'll do that.

jleclanche commented on djstripe/models/payment_methods.py:
```diff
@@ -84,18 +86,18 @@ class LegacySourceMixin:
-                "{}s must be manipulated through a Customer. "
-                "Pass a Customer object into this call.".format(cls.__name__)
-            )
+    # @classmethod
```
You're good to delete stuff in your commits rather than comment them out.

arnav13081994 replied:
Outdated

jleclanche commented on djstripe/models/payment_methods.py:
```diff
@@ -1,8 +1,10 @@
 import stripe
 from django.db import models, transaction
 from stripe.error import InvalidRequestError
+from stripe.six import reraise
```
Any reason you're using six here instead of native reraising?

arnav13081994 replied:
Outdated. Not using six or that complex error-handling and `re-raising` error approach in the final piece of code.

jleclanche commented on djstripe/models/payment_methods.py:
```diff
@@ -127,39 +129,79 @@ def _get_customer_or_account_from_kwargs(cls, **kwargs):
+                )
+            # re-raise error
+            if saved_error:
+                reraise(saved_error[0], saved_error[1], saved_error[2])
```
Not a fan of this pattern

arnav13081994 replied:
Outdated. In the final piece of code, I adopted a much more simple and readable approach. This approach got out of hand, I agree.

jleclanche reviewed (changes requested):
I got some time to do a rough first pass over the commits. Looks very good overall, couple things to address.

arnav13081994 committed:
Updated djstripe_sync_models management command to handle syncing BankAccount and Card objects associated with Stripe Express and Custom Connected Accounts

arnav13081994 committed:
Added support for account.external_account.* and account.updated Stripe Connect Webhooks

This includes all CrudType events for when the connected account adds a Bank Account or a Card object as an external_account Payout Source. Stripe Custom Accounts can specify Cards as Payout sources.

Also updated _api_create(), api_list(), and api_retrieve() methods to first query by the customer model field (if present) and only then query by the account model field. This allows handling of cases where both fields may be present on a Card and/or BankAccount object. Handles the  corner case for marketplaces that allow a seller to also function as a buyer using the same Payment Source as their external connected account payout source.

arnav13081994 committed:
Added fixtures simulating Stripe webhook events: account.external_account.*, account.application.* and account.updated

Added fixtures for Stripe Standard, Custom, and Express Accounts

Swapped default_account() with FAKE_STANDARD_ACCOUNT.create() to better replicate Stripe's behaviour.

Removed redundant and unused fixtures.

arnav13081994 committed:
Added tests for all Stripe account.* webhook events for all 3 types of Stripe Connected Accounts

Also updated test modules - test_card and test_bankaccount -  to ensure correct error messages get emitted in case neither the customer nor account model fields are present or the fields are not of the expected class types.
